libparsec/crates/types/src/id.rs

```diff
@@ -675,18 +668,133 @@ impl TryFrom<(&str, &str)> for HumanHandle {
     type Error = HumanHandleParseError;
 
     fn try_from((email, label): (&str, &str)) -> Result<Self, Self::Error> {
-        Self::new(email, label)
+        Self::from_raw(email, label)
     }
 }
 
 impl From<HumanHandle> for (String, String) {
     fn from(item: HumanHandle) -> (String, String) {
-        (item.email, item.label)
+        (item.email.to_string(), item.label)
     }
 }
 
 crate::impl_from_maybe!(Option<HumanHandle>);
 
+#[derive(Clone, SerializeDisplay, DeserializeFromStr)]
+pub struct EmailAddress {
+    value: email_address_parser::EmailAddress,
+    display: String,
+}
+
+impl EmailAddress {
+    fn new(email: email_address_parser::EmailAddress) -> Self {
+        let display = email.to_string();
+        Self {
+            value: email,
+            display,
+        }
+    }
+}
+
+impl PartialEq for EmailAddress {
+    fn eq(&self, other: &Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl Eq for EmailAddress {}
+
+impl Hash for EmailAddress {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.value.hash(state);
+    }
+}
+
+impl Deref for EmailAddress {
+    type Target = email_address_parser::EmailAddress;
+
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+impl AsRef<str> for EmailAddress {
+    fn as_ref(&self) -> &str {
+        &self.display
+    }
+}
+
+impl PartialOrd for EmailAddress {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for EmailAddress {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.get_domain()
+            .cmp(other.get_domain())
+            .then_with(|| self.get_local_part().cmp(other.get_local_part()))
+    }
+}
+
+impl std::fmt::Debug for EmailAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(&self.value, f)
+    }
+}
+
+impl Display for EmailAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.display)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum EmailAddressParseError {
+    #[error("Failed to parse email: {}", .0)]
+    ParseError(#[from] ::core::fmt::Error),
+    #[error("Invalid email domain")]
+    InvalidDomain,
+}
+
+impl FromStr for EmailAddress {
+    type Err = EmailAddressParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // A word about `<string>.nfc()`: In the unicode code we have multiple forms to represent the same glyph.
+        // We have 2 notable forms _Normalization From canonical Decomposition_ (NFD) and _Normalization From canonical Composition_ (NFC)
+        // For example: the `small letter A with acute` (á) would be encoded in NFD as `small letter a + acute accent` as for NFC `small letter a with acute`.
+        //
+        // So we need to normalize the string to have consistant comparison latter on.
```

Comment:
:+1: for the comment. A couple of suggestions and fixes: 
```suggestion
        // A word about `<string>.nfc()`: 
        // In Unicode there may be multiple ways to represent the same character. 
        // The Unicode standard defines Normalization Forms (such as NFC and NFD) 
        // to be able to determine if two Unicode strings are equivalent.
        // For example: the `small letter A with acute` (á) would be encoded in NFD
        // as `small letter a + acute accent` and in NFC as `small letter a with acute`.
        // See: https://www.unicode.org/reports/tr15/#Norm_Forms
        //
        // So we need to normalize the string to have consistent comparison later on.
```